### ISSUE TITLE
Helper functions don't perform type narrowing operations

### ISSUE THREAD

**Issue by MiguelGuthridge** [python/mypy#12148]
**Bug Report**

Sometimes, a function needs to check things already checked in another function in order to type narrow a variable. It's bad code design to copy the code, but if I use said function to type narrow a variable, MyPy doesn't detect this, resulting in incorrect errors. Although the example here is simple, more complex cases could result in huge amounts of repeated code without the use of helper functions like this.

**To Reproduce**

The following code snippet reproduces the error:
```py
from typing import Optional

def checkThing(thing: Optional[int]) -> bool:
    return thing is not None

a: Optional[int] = 3

if checkThing(a):
    print(a + 3) # Error: MyPy thinks that `a` could still be `None`
```

**Expected Behavior**

MyPy would infer from the function call to `checkThing()` that `a` cannot be None (otherwise we'd never enter the if block).

**Actual Behavior**

MyPy doesn't narrow any information on the type of the variable.

**Your Environment**

- Mypy version used: mypy 0.931
- Mypy command-line flags: [run by MyPy VS Code extension with default settings]
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: Python 3.9.5
- Operating system and version: Windows 11

Apologies if this is a duplicate - I searched for it but couldn't find anything matching.

**Comment by MiguelGuthridge** [comment]
Turns out that the reason I didn't find a solution was because Python 3.9. Just need to use `TypeGuard` from typing_extensions.
